### PR TITLE
[mlir][vector] Add tests for xfer-permute-lowering (1/n)(nfc)

### PR DESCRIPTION
**Pull request by banach-space**
Adds more tests to "vector-transfer-permutation-lowering.mlir",
specifically for the `TransferWritePermutationLowering` pattern - such
tests seem to be missing ATM.

The following edge cases are covered:
  * plain fixed-width (supported)
  * scalable vectors with mask (supported)
  * plain fixed-width, masked (not supported)

This is a part of a larger effort to make sure that all key cases for
patterns under `populateVectorTransferPermutationMapLoweringPatterns`
(*) are tested. I also want to make sure that tests use consistent
function and variable names.

(*) `transform.apply_patterns.vector.transfer_permutation_patterns` in
TD parlance)

### CODE REVIEW

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 1]
```diff
@@ -1,14 +1,81 @@
...
-/// Input: 
-///   * vector.transfer_write op with a map which _is not_ the permutation of a
-///     minor identity
+/// Input:
+///   * vector.transfer_write op with a permutation that under a transpose
+///     _would be_ a permutation of a minor identity
 /// Output:
-///   * vector.broadcast + vector.transfer_write with a map which _is_ the permutation of a
+///   * vector.transpose + vector.transfer_write with a map which _is_ a
+///     permutation of a minor identity
```
Comment (nujaa) [inline mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:11]: It is not a permutation of a minor identity but a minor identity per se.
See examples in 
https://github.com/llvm/llvm-project/blob/77db8b08c8b186c2625f8dfb26bb976561b43c4c/mlir/lib/Dialect/Vector/Transforms/LowerVectorTransfer.cpp#L156-L174
outputs are all of identity like.  The goal of those patterns is to generate identity permutation map.

**Review by nujaa — CHANGES_REQUESTED** [review]
Hi, thanks, it's a good start, I think there is a misunderstanding on the scope of those patterns. Feel free to tell me I'm wrong.

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir (hunk 1 above)
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:11]: That's a very good point, thanks!

I read this comment a while back and just have it hard-coded in my head that it's "the permutation of a minor identiy":
https://github.com/llvm/llvm-project/blob/77db8b08c8b186c2625f8dfb26bb976561b43c4c/mlir/lib/Dialect/Vector/Transforms/LowerVectorTransfer.cpp#L231-L245

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 2]
```diff
@@ -1,14 +1,81 @@
...
 ///----------------------------------------------------------------------------------------
-/// vector.transfer_write
+/// vector.transfer_write -> vector.transpose + vector.transfer_read
 ///----------------------------------------------------------------------------------------
-/// Input: 
-///   * vector.transfer_write op with a map which _is not_ the permutation of a
-///     minor identity
+/// Input:
+///   * vector.transfer_write op with a permutation that under a transpose
+///     _would be_ a permutation of a minor identity
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:8]: Let me rephrase and tell me what you think.

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 3]
```diff
@@ -1,14 +1,81 @@
 // RUN: mlir-opt %s --transform-interpreter --split-input-file | FileCheck %s
 
 ///----------------------------------------------------------------------------------------
-/// vector.transfer_write
+/// vector.transfer_write -> vector.transpose + vector.transfer_read
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:4]: > Maybe naming the actual pattern being tested here TransferWritePermutationLowering makes sense.

This is a bit tricky - what do you we for tests that are meant to trigger multiple patterns? What if later such patterns are split into other patterns? I guess it's safe if it's just one pattern, but doesn't scale well.

> Overall, I find the name of this pattern maybe a bit confusing. 

* It's a pattern for `vector.transfer_write`, hence `TransferWrite`
* It looks at the attached `permutation_map` to decide what to do, hence `Permutation`
* It "lowers" a complex `xfer_write` with an explicit permutation map, into a much simpler `xfer_write`

Also, from https://mlir.llvm.org/docs/Dialects/Vector/#vectortransfer_read-vectortransferreadop,
> The permutation_map [attribute](https://mlir.llvm.org/docs/LangRef/) is an [affine-map](https://mlir.llvm.org/docs/Dialects/Affine/) which specifies the transposition on the slice to match the vector shape. The permutation map may be implicit and omitted from parsing and printing if it is the canonical minor identity map (i.e. if it does not permute or broadcast any dimension).

So a minor identity map is consider the "lower level" canonical form. In this sense then name makes sense. WDYT? I'm happy to rename it though if there's sth better.

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 4]
```diff
@@ -1,14 +1,81 @@
...
 /// Output:
-///   * vector.broadcast + vector.transfer_write with a map which _is_ the permutation of a
+///   * vector.transpose + vector.transfer_write with a map which _is_ a
+///     permutation of a minor identity
+
+// CHECK-LABEL:   func.func @xfer_write_perm_minor_id_with_transpose(
+// CHECK-SAME:       %[[ARG_0:.*]]: vector<4x8xi16>,
+// CHECK-SAME:       %[[MEM:.*]]: memref<2x2x8x4xi16>) {
+// CHECK:           %[[TR:.*]] = vector.transpose %[[ARG_0]], [1, 0] : vector<4x8xi16> to vector<8x4xi16>
+// CHECK:           vector.transfer_write %[[TR]], %[[MEM]]{{.*}} {in_bounds = [true, true]} : vector<8x4xi16>, memref<2x2x8x4xi16>
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:17]: There is no permutation_map, it becomes implicit ;-) Let make add an explicit check for that.

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 5]
```diff
@@ -1,14 +1,81 @@
...
+    permutation_map = affine_map<(d0, d1, d2, d3) -> (d3, d2)>
+    } : vector<4x8xi16>, memref<2x2x8x4xi16>
+  } : vector<8x4xi1>
+
+  return
+}
+
+///----------------------------------------------------------------------------------------
+/// vector.transfer_write -> vector.broadcast + vector.transpose + vector.transfer_read
+///----------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:72]: > Do you think we could find a test case isolating patterns ?

That's going to be hard given how `populateVectorTransferPermutationMapLoweringPatterns` is defined. I also think that it's fine to keep the granularity as is. At least for now.

In general, I agree that it would be nice to be able to test every pattern in isolation. But then we would find ourselves testing all possible combination as well (in form of "integration tests"). That could be too much 😅 IMHO, we should be thinking in terms of code coverage instead. Would splitting patterns help here?

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 6]
```diff
@@ -1,14 +1,81 @@
...
+/// Input:
+///   * vector.transfer_write op with a map which _is not_ a permutation of a
 ///     minor identity
+/// Output:
+///   * vector.broadcast + vector.transpose + vector.transfer_write with a map
+///     which _is_ a permutation of a minor identity
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:78]: Will fix this in a follow-up 👍🏻

File: mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir [hunk 7]
```diff
@@ -1,14 +1,86 @@
...
+///     _is_ a minor identity
+
+// CHECK-LABEL:   func.func @xfer_write_transposing_permutation_map
+// CHECK-SAME:       %[[ARG_0:.*]]: vector<4x8xi16>,
+// CHECK-SAME:       %[[MEM:.*]]: memref<2x2x8x4xi16>) {
+// CHECK:           %[[TR:.*]] = vector.transpose %[[ARG_0]], [1, 0] : vector<4x8xi16> to vector<8x4xi16>
+// CHECK:           vector.transfer_write
+// CHECK-NOT:       permutation_map
+// CHECK-SAME:      %[[TR]], %[[MEM]]{{.*}} {in_bounds = [true, true]} : vector<8x4xi16>, memref<2x2x8x4xi16>
+func.func @xfer_write_transposing_permutation_map
```
Comment (nujaa) [inline mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir:21]:
```suggestion
func.func @xfer_write_transposing_permutation_map(
```
Missing `(`

**Review by nujaa — COMMENTED** [review]
Good points, I think the pipeline fails because of that missed parenthesis. Else, LGTM.

**Review by nujaa — APPROVED** [review]
